src/devtools/server.ts

```diff
@@ -15,8 +15,53 @@ devtoolsEvents.addListener("fromDevTools", (msg) => {
   Debug.info("DevTools > WS: " + JSON.stringify(msg));
 });
 
-export function startServer(port: number): Disposable {
+let id = 1;
+
+export function sendToDevTools(message: unknown) {
+  devtoolsEvents.emit("toDevTools", {
+    id: `vscode-${id++}`,
+    source: "apollo-client-devtools",
+    type: "actor",
+    message,
+  });
+}
+
+export let serverState:
+  | { port: false | number; disposable: Disposable }
+  | undefined = undefined;
+
+export function startServer(port: number) {
```

Comment:
A bunch of checks moved over from `extension.ts` here and the whole "disposing" and "restarting" should be a lot more solid now.